src/prev_content_serde.rs

```diff
@@ -0,0 +1,67 @@
+//! Serialize all event prev_content types when inside one of `StateEvent<C>`.
+//!
+//! `prev_content` is always `Option<C>`.
+
+use std::any::type_name;
+use std::fmt;
+use std::marker::PhantomData;
+
+use serde::{
+    de::{self, Deserialize, Deserializer, Error as _, MapAccess, SeqAccess, Visitor},
+    ser::SerializeStruct,
+    Serializer,
+};
+use serde_json::value::to_raw_value;
+
+use crate::EventContent;
+
+/// Serialize event content types when wrapped in `Option`.
+///
+/// Used when serializing the prev_content field of `StateEvent` types like
+/// so `#[serde(with = "crate::prev_content_type_serde")]`.
+pub fn serialize<T, S>(content: &Option<T>, serializer: S) -> Result<S::Ok, S::Error>
+where
+    T: EventContent,
+    S: Serializer,
+{
+    // this only runs when Option::Some
+    if content.is_some() {
+        let mut state = serializer.serialize_struct(type_name::<T>(), 1)?;
+        state.serialize_field("prev_content", &content)?;
+        state.end()
+    } else {
+        serializer.serialize_none()
+    }
+}
+
+/// Deserialize event content types when wrapped in `Option`.
+///
+/// Used when serializing the prev_content field of `StateEvent` types like
+/// so `#[serde(with = "crate::prev_content_type_serde")]`.
+pub fn deserialize<'de, T, D>(deserializer: D) -> Result<Option<T>, D::Error>
+where
+    T: EventContent,
+    D: serde::de::Deserializer<'de>,
+{
+    #[derive(serde::Deserialize)]
+    struct Event {
+        #[serde(flatten)]
+        content: serde_json::Value,
```

Comment:
Once the general flatten issue is resolved, this should be

```suggestion
        prev_content: Option<serde_json::Value>,
```